src/absil/ilreflect.fs

```diff
@@ -1845,78 +1858,100 @@ let rec buildTypeDefPass3 cenv nesting modB emEnv (tdef : ILTypeDef) =
 // nested type by calling CreateType on the TypeBuilder object that represents 
 // the nested type. The code example for this topic shows how to define such 
 // an event handler.
+//
+//  
+// There is also a case where generic parameter constraints were being checked before
+// a generic method was called.  This forced the loading of the types involved in the 
+// constraints very early.
+//
+
 //----------------------------------------------------------------------------
 
 let getEnclosingTypeRefs (tref:ILTypeRef) = 
    match tref.Enclosing with 
    | [] -> []
    | h :: t -> List.scan (fun tr nm -> mkILTyRefInTyRef (tr,nm)) (mkILTyRef(tref.Scope, h)) t
 
-let rec getTypeRefsInType valueTypesOnly typ acc = 
+[<RequireQualifiedAccess>]
+type AllTypes = Yes | No
+
+// Find all constituent type references
+let rec getTypeRefsInType (allTypes: AllTypes) typ acc = 
     match typ with
-    | ILType.Void | ILType.TypeVar _                              -> acc
-    | ILType.Ptr eltType | ILType.Byref eltType -> getTypeRefsInType valueTypesOnly eltType acc
-    | ILType.Array (_,eltType) -> if valueTypesOnly then acc else getTypeRefsInType valueTypesOnly eltType acc
-    | ILType.Value tspec -> tspec.TypeRef :: List.foldBack (getTypeRefsInType valueTypesOnly) tspec.GenericArgs acc
-    | ILType.Boxed tspec -> if valueTypesOnly then acc else tspec.TypeRef :: List.foldBack (getTypeRefsInType valueTypesOnly) tspec.GenericArgs acc
+    | ILType.Void 
+    | ILType.TypeVar _  -> acc
+    | ILType.Ptr eltType | ILType.Byref eltType -> 
+        getTypeRefsInType allTypes eltType acc
+    | ILType.Array (_,eltType) -> 
+        match allTypes with 
+        | AllTypes.No -> acc 
```

Comment:
Before it was clear that it applied to ValueTypesOnly from the naming.  It's not clear what distinguishes between alltypes yes and alltypes no?